### PR TITLE
docs: add documentation for model-hub [DET-5031]

### PR DESCRIPTION
**Pull request by liamcli**
## Checklist

- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.
- [x] Licenses should be included for new code which was copied and/or modified from any external code.

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

### CODE REVIEW

File: docs/model-hub/index.txt [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+###########
+
+**********
+ Overview
+**********
+
+Determined's **model-hub** library makes it easy to train models from popular
+third-party libraries with a Determined cluster. With model-hub, use trusted
+implemenations of model architectures, with Determined's ability to scale
+training, track experiments, share resources and perform hyperparameter
```
Comment (naren-determined) [inline docs/model-hub/index.txt:14]: I think we should use the term distributed training instead of scale training, but not super strongly opinionated there.

Nit: Add comma `share resources, and perform...`

File: docs/model-hub/index.txt [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+-  Official examples checked for correctness and thoroughly tested for
+   use with Determined.
+-  A base Determined Trial class with common functionality implemented
+   for the user.
+-  A prebuilt docker environment with all dependencies installed and
+   versioned for reproducibility.
+-  A suite of helper functions (if applicable) to allow users to easily
+   write their own trial classes for use with the third-party library.
+
+**********
```
Comment (naren-determined) [inline docs/model-hub/index.txt:28]: I think the linter gets mad if the **** don't exceed the text length

File: docs/model-hub/index.txt [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+<model-hub-transformers>` the `Huggingface transformers library
+<https://github.com/huggingface/transformers>`_.
+
+Future libraries on our roadmap to include in model-hub include
+
+-  `mmdetection <https://github.com/open-mmlab/mmdetection>`_
+-  `detectron2 <https://github.com/facebookresearch/detectron2>`_
+
+If you have additional libraries you want to see supported in
+**model-hub** please let us know!
```
Comment (naren-determined) [inline docs/model-hub/index.txt:62]: Maybe link to community Slack?

File: docs/model-hub/transformers/examples.txt [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+fact, each `transformer example corresponding to a core transformer task
+<https://huggingface.co/transformers/examples.html#the-big-table-of-tasks>`_
+has an associated task in **model-hub** that is guaranteed to work with
+Determined and verified for correctness. See the table below for a
+summary of the **model-hub** transformers examples:
+
+.. list-table::
+   :header-rows: 1
+
+   -  -  Task
```
Comment (naren-determined) [inline docs/model-hub/transformers/examples.txt:21]: I would rather have Task, Dataset, Filename like in the existing https://docs.determined.ai/latest/examples.html. I don't think the Description field really adds much here, and like the idea of a tarball rather than specific links, but not sure if other folks' opinions here.

File: model_hub/examples/huggingface/question-answering/qa_trial.py [hunk 5]
```diff
@@ -166,13 +166,13 @@ def build_validation_data_loader(self) -> det_torch.DataLoader:
     def evaluate_batch(self, batch: det_torch.TorchData, batch_idx: int) -> Dict:
```
Comment (naren-determined) [inline model_hub/examples/huggingface/question-answering/qa_trial.py:166]: Can you put these changes in the other examples PR -- just want to separate out docs from example fixes.

**Review by naren-determined — CHANGES_REQUESTED** [review]
tutorial.txt is awesome I love it!! Want to take another pass after these changes, but overall I like the structure of it.

**Comment by neilconway** [comment]
When I try to build this PR, I get

```
$ make # from top-level of the repo
[...]
Running Sphinx v2.4.4
loading translations [en]... done
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 122 source files that are out of date
updating environment: [new config] 122 added, 0 changed, 0 removed
2021-05-05 15:17:55.603179: W tensorflow/stream_executor/platform/default/dso_loader.cc:60] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory
2021-05-05 15:17:55.603219: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
/home/ubuntu/.envs/det/lib/python3.7/site-packages/torch/cuda/__init__.py:52: UserWarning: CUDA initialization: Found no NVIDIA driver on your system. Please check that you have an NVIDIA GPU and installed a driver from http://www.nvidia.com/Download/index.aspx (Triggered internally at  /pytorch/c10/cuda/CUDAFunctions.cpp:100.)
  return torch._C._cuda_getDeviceCount() > 0
reading sources... [100%] tutorials/tf-mnist-tutorial
sphinx-sitemap warning: No pages generated for sitemap.xml

Warning, treated as error:
autodoc: failed to import class 'huggingface.BaseTransformerTrial' from module 'model_hub'; the following exception was raised:
No module named 'model_hub'
```

File: docs/model-hub/transformers/api.txt [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+   :pyobject: BaseTransformerTrial.train_batch
+
+.. _transformers-functions:
+
+Helper Functions
+================
+
+The ``BaseTransformerTrial`` calls many helper functions below that are
+also useful for those subclassing ``BaseTransformerTrial`` or writing
+their own transformers trials for use with Determined.
```
Comment (naren-determined) [inline docs/model-hub/transformers/api.txt:37]: Nit: Their --> your

File: docs/model-hub/transformers/api.txt [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+.. automodule:: model_hub.huggingface
+   :members: default_parse_config_tokenizer_model_kwargs, default_parse_optimizer_lr_scheduler_kwargs, build_using_auto, build_default_optimizer, build_default_lr_scheduler, default_load_dataset
+
+Structured Dataclasses
+======================
+
+Structured dataclasses are used to ensure that we parse the experiment
```
Comment (naren-determined) [inline docs/model-hub/transformers/api.txt:45]: Nit: we parse --> Determined parses

File: docs/model-hub/transformers/index.txt [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+
+########################
+ Model Hub Transformers
+########################
+
+.. rubric:: **Overview**
+
+`The Huggingface transformers library
+<https://github.com/huggingface/transformers>`_ is the de-facto library
+for natural language processing models. It provides pretrained weights
```
Comment (naren-determined) [inline docs/model-hub/transformers/index.txt:11]: Might want to put the acronym here since you use it later. i.e. "for natural language processing (NLP) models"

File: docs/model-hub/transformers/index.txt [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+to what it's like working directly with **transformers**. Our library
+serves as an alternative to HuggingFace's `Trainer Class
+<https://huggingface.co/transformers/main_classes/trainer.html>`_ and
+provides access all of Determined's benefits including:
+
+-  Easy :ref:`multi-node distributed training <multi-gpu-training>` with
+   no code modifications. Determined automatically sets up the
+   distributed backend for you.
+
+-  Experiment tracking, experiment visualization, artifact tracking, and
```
Comment (naren-determined) [inline docs/model-hub/transformers/index.txt:27]: Experiment visualization as a thing doesn't make too much sense to me. What specifically are you visualizing here -- the loss accuracy curves or the hyperparameters? Dunno if I'm being too nitpicky on this so feel free to ignore and resolve this comment if I am.

File: docs/model-hub/transformers/index.txt [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+   :ref:`state-of-the-art hyperparameter search <hyperparameter-tuning>`
+   without requiring third-party integrations.
+
+-  :ref:`Automated cluster management, fault tolerance, and job
+   rescheduling <benefits-of-determined>` so you don't have to worry
+   about provisioning resources or babysit your experiments.
```
Comment (naren-determined) [inline docs/model-hub/transformers/index.txt:33]: Nit: babysit --> babysitting

**Review by naren-determined — COMMENTED** [review]
Overall looks good -- once these comments are addressed I would be okay merging.

File: docs/model-hub/transformers/tutorial.txt [hunk 11]
```diff
@@ -0,0 +1,297 @@
...
+   :py:class:`model_hub.huggingface.BaseTransformerTrial` parent
+   ``__init__`` function to customize how the **transformers** config,
+   tokenizer, and model objects are constructed.
+
+Let us walk through qa_trial.py_ to learn the basics. We won't cover the
+model definition line by line but will highlight the parts that make use
+of **model-hub**.
+
+After this tutorial, if you want to further customize a trial for your
+own use, you can look at qa_beam_search_trial.py_ for an example.
```
[the remark this answers is not in the document]
Comment (liamcli) [reply docs/model-hub/transformers/tutorial.txt]: the underscore tells sphinx that it's referencing the defined variable in line 13 which links to the actual .py file.

**Review by naren-determined — APPROVED** [review]
Lgtm 🚢
